Adding support for Ubunt 16.04 and Centos 7. Breaking changes

hyeomans opened the pull request:
Looks like mcrouter has changed a lot since 2015, however I tried to make minimal changes.

* Added support for Ubuntu 16.04 and CentOS 7
* I could not get mcrouter to compile on Ubuntu 14.04, at least with the latest version
* New libraries have to built before building mcrouter, I added those (double_conversion and wangle)
* I modified the way to get the latest version for mcrouter. I could not get the `git` resource to get the latest commit from the latest release branch.
* Tests need to be improved, but I will like to get that on a follow up PR.
* systemd is used by default

Overall the whole build process had to be redone but if you take a closer look, starting and passing variables to the service wasn't modified.

jeffbyrnes commented on recipes/default.rb:
```diff
@@ -16,9 +17,13 @@
 include_recipe 'mcrouter::install'
 include_recipe 'mcrouter::configure'
-include_recipe 'mcrouter::service'
+include_recipe 'mcrouter::systemd'
```
I think it’s preferable to leave this recipe as `::service`, since it _remains_ a service provider. That way, if we shift to another one in future, it’s not yet another name change.

Also, we may want to provide additional service systems.

hyeomans replied:
Got it.

jeffbyrnes commented on recipes/install.rb:
```diff
@@ -16,40 +17,39 @@
+execute 'install mcrouter' do
+  cwd mcrouter_build_dir
+  command install_command
+  not_if { ::File.exist?(mcrouter_bin) }
```
Since `execute` has the `creates` functionality built in, there’s no need to use a guard like this; it works the same way (and is easier to parse, to boot).

hyeomans replied:
Great, I should have said I'm kind of new into Chef, thanks for the great feedback.

jeffbyrnes replied:
Gotcha!

Yeah, it may help to realize that the whole odd-seeming model of all those resources with `action :nothing` is b/c `execute` resources are not idempotent by default (they can end up with different results each time).

I may do some modifications to all this code to bring it back to a scheme that better aligns with Chef & the need for idempotency, but this is a good place to jump off from.

jeffbyrnes commented on spec/spec_helper.rb:
```diff
@@ -1,14 +1,10 @@
 require 'chefspec/berkshelf'
-require 'coveralls'
 
-Coveralls.wear!
```
Why the removal of the Coveralls code?

hyeomans replied:
It was broken and could not get it up to get it working again. Maybe something change?

jeffbyrnes replied:
Yeah, I’m cleaning up the testing code now myself. I see what you mean! I’ll sort it out; please ignore this comment for now.

hyeomans commented on CHANGELOG.md:
```diff
@@ -1,5 +1,16 @@
+* Tests need to be improved
+
+### Changes:
+* Added support for Ubuntu 16.04, CentOS 7.2 and CentOS 7.3
```
yeah, sorry about that one. When testing it locally, kitchen ran out of memory if I tried to converge both ubuntu and centos, will fix.

hyeomans commented on metadata.rb:
```diff
@@ -1,15 +1,23 @@
-maintainer       'EverTrue, Inc.'
-maintainer_email 'devops@evertrue.com'
+maintainer       'EverTrue, Inc., Roblox Inc.'
+maintainer_email 'devops@evertrue.com, devops@roblox.com'
```
Got it, no worries.

hyeomans commented on attributes/default.rb:
```diff
@@ -1,29 +1,28 @@
     'A' => {
       'servers' => [
-        'localhost:11811'
+        '0.0.0.0:11811'
```
When I was testing locally in kitchen, I could not it get it working on 127.0.0.1, I'll keep testing

jeffbyrnes replied:
Hmm. That’s rather odd, but makes sense. I’ll check it out myself!

hyeomans commented:
Regarding Folly, even facebook builds it in this weird way where they hard code it into the repo:
https://github.com/facebook/mcrouter/blob/release-36-0/mcrouter/FOLLY_COMMIT
 and read it through their build scripts:
https://github.com/facebook/mcrouter/blob/master/mcrouter/scripts/recipes/folly.sh#L8

Which the default is to have the HEAD of master but that doesn't work. 

If you download the release package, you don't get those files, FOLLY_COMMIT and WANGLE_COMMMIT, which kind of defeats the purpose of having `ark`. I'm not saying I want to remove it, it just made sense given how mcrouter project is structured. Looks like there is no correct way to build mcrouter 🤔

jeffbyrnes commented:
@hyeomans that is… very odd. Thanks for tracking down how their releases work! I’ll look into it, as well, but if we need to use the `git` resource, so be it.

hyeomans commented:
hey @jeffbyrnes I updated addressing almost all comments, I'm missing testing changing from `/usr/local` to `/opt/mcrouter`. 

Also updated the README.md file with more information, hopefully it makes a little bit of more sense.

jeffbyrnes commented:
Thanks @hyeomans! I’ve gotta take a break from this, as I’ve to finish some Real Work™ things today, and am on holiday until Tuesday, so I’ll return to this next week. I hope that’s alright.

My focus, when I get back, will be to reintroduce testing to this, and see if we can avoid diverging quite so much from the previous model of the cookbook.

I did grok what you were explaining w/ Facebook’s odd release mechanism: they tag a commit (e.g., `v0.36.0`), then create a release branch from that commit and add the `FOLLY_COMMIT` and `WANGLE_COMMIT` files. So perhaps we can just grab an archive of a particular branch, and avoid doing all manner of git things.

jeffbyrnes commented:
I’m going to try & carve out some time this week to work on this.

jeffbyrnes commented on recipes/_folly.rb:
```diff
@@ -0,0 +1,48 @@
+  action :delete
+  recursive true
+  only_if { Dir.exist?(folly_checkout_path) }
+end
```
I’d discourage doing this, since it breaks the idempotency of this recipe. After the first convergence, this is what will happen with this:

1. The `folly_checkout_path` will be deleted (i.e., the git clone, which contains the build dir)
2. The git repo will be cloned again

There’s no need to remove the git repo, only to re-create it every time.

After thinking on this some more, the best pattern here is probably to make the `execute` resources `action :nothing`, and notify them from the git repo. That means, on any change to the repo code, the library is rebuilt and reinstalled.

I’m gonna see about cleaning that up myself, since I have some other changes I want to make.
